README.md
# PHP-IPAddress

A set of utility classes for working with IP addresses in PHP.
Supports both IPv4 and IPv6 schemes.

Fork to support PHP 8.1

### Requirements

* PHP version 8.1 or greater.
* The [PEAR](http://pear.php.net/) [Math_BigInteger](http://pear.php.net/package/Math_BigInteger/) class

  Required for add & subtract operations on IPv6 addresses, and finding IPs in IPv6 address blocks.

## Examples

```php
<?php
use Leth\IPAddress\IP, Leth\IPAddress\IPv4, Leth\IPAddress\IPv6;
// Creates an instance
$ip = IP\Address::factory('127.0.0.1');

// Prints "127.0.0.1"
echo $ip . "\n";

/**
 * IP\Address::factory(...) will attempt to guess the address version from the arguments
 */
// Returns an instance of IPv4\Address
$ip = IP\Address::factory('127.0.0.1');
$ip = IPv4\Address::factory('127.0.0.1');

// Returns an instance of IPv6\Address
$ip = IP\Address::factory('::1');
$ip = IPv6\Address::factory('::1');

/**
 * IP\NetworkAddress::factory(...) will also attempt to guess protocol versions
 */
// Can either be called with the subnet size encoded in the address string,
$net_addr = IP\NetworkAddress::factory('192.168.0.1/24');
// Or as the second parameter
$net_addr = IP\NetworkAddress::factory('192.168.0.1', 24);

// Prints '192.168.0.0'
echo $net_addr->get_network_address() . "\n";
// Prints '192.168.0.255'
echo $net_addr->get_broadcast_address() . "\n";
// Prints '255.255.255.0'
echo $net_addr->get_subnet_mask() . "\n";

/**
 * For each address of the specified network.
 */
$network = IPv4\NetworkAddress::factory('192.168.0.0/24');
foreach ($network as $ip) {
	// $ip is instance of IPv4\Address with value:
	// 192.168.0.0
	// 192.168.0.1
	// ...
}

$network = IPv4\NetworkAddress::factory('192.168.0.0/24');
// Prints '256'
echo count($network);

/**
 * Merge adjacent NetworkAddress blocks into larger blocks
 */
$small = [
	IPv4\NetworkAddress::factory('192.168.0.0/24'),
	IPv4\NetworkAddress::factory('192.168.1.0/24')
];
$merged = IP\NetworkAddress::merge($small);
// Prints '1'
echo count($merged);
// Prints '1'
echo $merged[0] == IP\NetworkAddress::factory('192.168.0.0/23');

/**
 * Get specified octet from IP
 */
$ipv4 = IP\Address::factory('192.168.1.102');
// Prints '102'
echo $ipv4->get_octet(-1);
// Prints '168'
echo $ipv4[1];

$ipv6 = IP\Address::factory('2490::fa');
// Prints '250'
echo $ipv6->get_octet(-1);
// Prints '0'
echo $ipv6[5];
```

## Test Cases

To run the test cases, the following commands will do the trick:

*	No-frills tests:

	phpunit -c phpunit.xml.dist

*	Generate code coverage reports into './coverage/':

	phpunit -c phpunit.xml.dist --coverage-html coverage

*	With colours and verbose output:

	phpunit -c phpunit.xml.dist --colors --verbose

*	All together:

	phpunit -c phpunit.xml.dist --coverage-html coverage --colors --verbose


CHANGELOG.md
# PHP-IPAddress

A set of utility classes for working with IP addresses in PHP.
Supports both IPv4 and IPv6 schemes.

Fork to support PHP 8.1

## Ver 2.0.0

PHP 8.1 required


## New features
is_same_version is now public
## Backwards compability breaking changes:

#### get_block_in_smallest()
 Now throws instead of returning [null, null] when no smallest block is found

#### get_octet
Now throws InvalidArgumentException instead of returning null when no smallest block is found


tests/IPAddressTest.php
<?php
use Leth\IPAddress\IP, Leth\IPAddress\IPv4, Leth\IPAddress\IPv6;
use PHPUnit\Framework\TestCase;

/**
 * Tests for the IP\Address Class
 *
 * @package default
 * @author Marcus Cobden
 */
class IP_Address_Test extends TestCase
{

	public function providerFactory()
	{
		return [
			['IPv4\\Address', '127.0.0.1',   '127.000.000.001',   '127.0.0.1'],
			['IPv4\\Address', IPv4\Address::factory('127.0.0.1'),   '127.000.000.001',   '127.0.0.1'],
			['IPv4\\Address', '127.0.0.0',   '127.000.000.000',   '127.0.0.0'],
			['IPv4\\Address', '127.0.0.2',   '127.000.000.002',   '127.0.0.2'],
			['IPv4\\Address', '192.168.1.1', '192.168.001.001', '192.168.1.1'],
			['IPv4\\Address', '192.168.2.1', '192.168.002.001', '192.168.2.1'],
			['IPv4\\Address', '192.168.1.2', '192.168.001.002', '192.168.1.2'],
			['IPv4\\Address', '10.0.0.2',    '010.000.000.002',    '10.0.0.2'],
			['IPv4\\Address', '10.0.0.1',    '010.000.000.001',    '10.0.0.1'],
			['IPv4\\Address', 257,           '000.000.001.001',     '0.0.1.1'],

			['IPv6\\Address', new Math_BigInteger(257),                  '0000:0000:0000:0000:0000:0000:0000:0101', '::101'],
			['IPv6\\Address',                                     '::1', '0000:0000:0000:0000:0000:0000:0000:0001', '::1'],
			['IPv6\\Address',              IPv6\Address::factory('::1'), '0000:0000:0000:0000:0000:0000:0000:0001', '::1'],
			['IPv6\\Address', '0000:0000:0000:0000:0000:0000:0000:0001', '0000:0000:0000:0000:0000:0000:0000:0001', '::1'],
			['IPv6\\Address',               'fe80::62fb:42ff:feeb:727c', 'fe80:0000:0000:0000:62fb:42ff:feeb:727c', 'fe80::62fb:42ff:feeb:727c'],
			['IPv6\\Address',               'fc00:1234::e0',             'fc00:1234:0000:0000:0000:0000:0000:00e0', 'fc00:1234::e0'],
		];
	}

	/**
	 * @dataProvider providerFactory
	 */
	public function testFactory($expected_class, $input, $full, $compact)
	{
		$expected_class = 'Leth\\IPAddress\\'.$expected_class;
		$instances = [IP\Address::factory($input), $expected_class::factory($input)];

		foreach ($instances as $instance) {
			$this->assertNotNull($instance);
			$this->assertEquals($full,    $instance->format(IP\Address::FORMAT_FULL));
			$this->assertEquals($compact, $instance->format(IP\Address::FORMAT_COMPACT));
			$this->assertEquals($expected_class, get_class($instance));
		}
	}

	public function providerFactoryException()
	{
		$data = [];
		$data[] = ['cake'];
		$data[] = ['12345'];
		$data[] = ['-12345'];
		return $data;
	}

	/**
	 * @dataProvider providerFactoryException
	 */
	public function testFactoryException($input)
	{
		$this->expectException(InvalidArgumentException::class);
		IP\Address::factory($input);
	}

	public function providerCompare()
	{
		$data = [];
		$data[] = ['127.0.0.1', '127.0.0.1', 0];
		$data[] = ['127.0.0.0', '127.0.0.1', -1];
		$data[] = ['127.0.0.0', '127.0.0.2', -1];
		$data[] = ['127.0.0.1', '127.0.0.2', -1];
		$data[] = ['127.0.0.2', '127.0.0.1', 1];
		$data[] = ['10.0.0.1',  '127.0.0.2', -1];
		return $data;
	}

	/**
	 * @dataProvider providerCompare
	 */
	public function testCompare($a, $b, $expected)
	{
		$result = IP\Address::compare(IP\Address::factory($a), IP\Address::factory($b));

		// Division is to ensure things are either -1, 0 or 1. abs() is to preseve sign.
		$this->assertEquals($expected, $result == 0 ? 0: $result / abs($result));
	}
}


tests/IPv4AddressTest.php
<?php

use Leth\IPAddress\IPv4, Leth\IPAddress\IPv6;
use Leth\IPAddress\IPv6\Address;
use PHPUnit\Framework\TestCase;

// class Testing_IPv4_Address extends IPv4\Address
// {
// 	public static function factory(mixed $address): IP\Address
// 	{
// 		return new Testing_IPv4\Address($address);
// 	}
// }

/**
 * Tests for the IPv4\Address Class
 *
 * @package default
 * @author Marcus Cobden
 */
class IPv4_Address_Test extends TestCase
{

	public function providerFactory()
	{
		$data = [];
		$data[] = ['127.0.0.1',  '127.0.0.1'];
		$data[] = ['127.0.0.0',  '127.0.0.0'];
		$data[] = ['127.0.0.2',  '127.0.0.2'];
		$data[] = ['192.168.1.1', '192.168.1.1'];
		$data[] = ['192.168.2.1', '192.168.2.1'];
		$data[] = ['192.168.1.2', '192.168.1.2'];
		$data[] = ['10.0.0.2',   '10.0.0.2'];
		$data[] = ['10.0.0.1',   '10.0.0.1'];
		$data[] = [new Math_BigInteger(1), '0.0.0.1'];
		$data[] = [new Math_BigInteger(2), '0.0.0.2'];
		$data[] = [new Math_BigInteger(3), '0.0.0.3'];
		$data[] = [new Math_BigInteger(256), '0.0.1.0'];
		return $data;
	}

	/**
	 * @dataProvider providerFactory
	 */
	public function testFactory($input, $expected)
	{
		$instance = IPv4\Address::factory($input);
		$this->assertNotNull($instance);
		$this->assertEquals($expected, (string) $instance);
	}

	public function testFormatInteger()
	{
		$ip = IPv4\Address::factory('127.0.0.1');
		$this->assertEquals(2130706433, $ip->format(IPv4\Address::FORMAT_INTEGER));
	}

	public function providerFormatException()
	{
		$bad_mode = -1;
		$data = static::providerFactory();
		foreach ($data as $i => $entry) {
			$data[$i] = [$entry[0], $bad_mode];
		}

		return $data;
	}

	/**
	 * @dataProvider providerFormatException
	 */
	public function testFormatException($input, $mode)
	{
		$this->expectException(InvalidArgumentException::class);
		$instance = IPv4\Address::factory($input);
		echo $instance->format($mode);
	}

	public function providerFactoryException()
	{
		$data = [];
		$data[] = ['256.0.0.1'];
		$data[] = ['127.-1.0.1'];
		$data[] = ['127.128.256.1'];
		$data[] = [new Math_BigInteger('99999999999999999')];
		$data[] = [123.45];
		$data[] = [-123.45];
		$data[] = ['cake'];
		$data[] = ['12345'];
		$data[] = ['-12345'];
		$data[] = ['0000:0000:0000:ffff:0127:0000:0000:0001'];
		return $data;
	}

	/**
	 * @dataProvider providerFactoryException
	 */
	public function testFactoryException($input)
	{
		$this->expectException(InvalidArgumentException::class);
		IPv4\Address::factory($input);
	}

	public function providerBitwise()
	{
		$data = [];
			//       OP1        OP2        AND        OR         XOR        NOT
		$data[] = ['0.0.0.1', '0.0.0.1', '0.0.0.1', '0.0.0.1', '0.0.0.0', '255.255.255.254'];
		$data[] = ['0.0.0.0', '0.0.0.1', '0.0.0.0', '0.0.0.1', '0.0.0.1', '255.255.255.255'];
		$data[] = ['0.0.0.1', '0.0.0.0', '0.0.0.0', '0.0.0.1', '0.0.0.1', '255.255.255.254'];
		$data[] = ['0.0.0.0', '0.0.0.0', '0.0.0.0', '0.0.0.0', '0.0.0.0', '255.255.255.255'];
		return $data;
	}

	/**
	 * @dataProvider providerBitwise
	 */
	public function testBitwise($ip1, $ip2, $ex_and, $ex_or, $ex_xor, $ex_not)
	{
		$ip1 = IPv4\Address::factory($ip1);
		$ip2 = IPv4\Address::factory($ip2);

		$this->assertEquals($ex_and, (string) $ip1->bitwise_and($ip2));
		$this->assertEquals($ex_or , (string) $ip1->bitwise_or($ip2));
		$this->assertEquals($ex_xor, (string) $ip1->bitwise_xor($ip2));
		$this->assertEquals($ex_not, (string) $ip1->bitwise_not());
	}

	// TODO Check this
	// public function providerAsIPv6Address()
	// {
	// 	return [
	// 		['127.0.0.1', '0000:0000:0000:ffff:0127:0000:0000:0001'],
	// 	];
	// }
	//
	// /**
	//  * @dataProvider providerAsIPv6Address
	//  */
	// public function testAsIPv6Address($v4, $v6)
	// {
	// 	$ip = IPv4\Address::factory($v4);
	//
	// 	$this->assertEquals($v6, (string) $ip->asIPv6Address());
	// }

	public function providerAddSubtract()
	{
		$data = [];
		$data[] = [IPv4\Address::factory('0.0.0.0')  , 0, IPv4\Address::factory('0.0.0.0')];
		$data[] = [IPv4\Address::factory('0.0.0.0')  , 1, IPv4\Address::factory('0.0.0.1')];
		$data[] = [IPv4\Address::factory('0.0.0.1')  , 0, IPv4\Address::factory('0.0.0.1')];
		$data[] = [IPv4\Address::factory('0.0.0.1')  , 1, IPv4\Address::factory('0.0.0.2')];
		$data[] = [IPv4\Address::factory('0.0.0.10') , 1, IPv4\Address::factory('0.0.0.11')];
		$data[] = [IPv4\Address::factory('0.0.0.255'), 1, IPv4\Address::factory('0.0.1.0')];
		$data[] = [IPv4\Address::factory('0.0.255.0'), 257, IPv4\Address::factory('0.1.0.1')];
		$data[] = [IPv4\Address::factory('255.255.0.0')  , 0, IPv4\Address::factory('255.255.0.0')];
		$data[] = [IPv4\Address::factory('255.255.0.0')  , 1, IPv4\Address::factory('255.255.0.1')];
		$data[] = [IPv4\Address::factory('255.255.0.1')  , 0, IPv4\Address::factory('255.255.0.1')];
		$data[] = [IPv4\Address::factory('255.255.0.1')  , 1, IPv4\Address::factory('255.255.0.2')];
		$data[] = [IPv4\Address::factory('255.255.0.10') , 1, IPv4\Address::factory('255.255.0.11')];
		$data[] = [IPv4\Address::factory('255.255.0.255'), 1, IPv4\Address::factory('255.255.1.0')];
		$data[] = [IPv4\Address::factory('255.0.255.0'), 257, IPv4\Address::factory('255.1.0.1')];
		$data[] = [IPv4\Address::factory('192.168.0.0'), 4, IPv4\Address::factory('192.168.0.4')];
		return $data;
	}

	/**
	 * @dataProvider providerAddSubtract
	 */
	public function testAddSubtract($left, $right, $expected)
	{
		$result = $left->add($right);
		$this->assertEquals(0, $result->compare_to($expected));
		$result = $result->subtract($right);
		$this->assertEquals(0, $result->compare_to($left));
	}

	public function providerAsIPv6Address()
	{
		$data = [];
		//                               input                               expected
		$data[] = [IPv4\Address::factory('0.0.0.0'  ), IPv6\Address::factory('::ffff:0:0'   )];
		$data[] = [IPv4\Address::factory('0.0.0.1'  ), IPv6\Address::factory('::ffff:0:1'   )];
		$data[] = [IPv4\Address::factory('0.0.0.255'), IPv6\Address::factory('::ffff:0:ff'  )];
		$data[] = [IPv4\Address::factory('0.0.255.0'), IPv6\Address::factory('::ffff:0:ff00')];
		$data[] = [IPv4\Address::factory('0.255.0.0'), IPv6\Address::factory('::ffff:ff:0'  )];
		$data[] = [IPv4\Address::factory('255.0.0.0'), IPv6\Address::factory('::ffff:ff00:0')];
		return $data;
	}

	/**
	 * @dataProvider providerAsIPv6Address
	 */
	public function testAsIPv6Address(IPv4\Address $input, IPv6\Address $expected_equal)
	{
		$converted = $input->as_IPv6_address();

		$this->assertInstanceOf(Address::class, $converted);
		$this->assertEquals(0, $converted->compare_to($expected_equal));
	}

	public function testGetOctet()
	{
		$ip = IPv4\Address::factory('10.250.30.40');

		$this->assertEquals(10, $ip->get_octet(-4));
		$this->assertEquals(250, $ip->get_octet(-3));
		$this->assertEquals(30, $ip->get_octet(-2));
		$this->assertEquals(40, $ip->get_octet(-1));

		$this->assertEquals(10, $ip->get_octet(0));
		$this->assertEquals(250, $ip->get_octet(1));
		$this->assertEquals(30, $ip->get_octet(2));
		$this->assertEquals(40, $ip->get_octet(3));
	}

	public function testGetOctetWithOutOfRangeOctet()
	{
		$ip = IPv4\Address::factory('10.250.30.40');
		$this->expectException(InvalidArgumentException::class);
		$ip->get_octet(4);
	}

	public function testArrayAccess()
	{
		$ip = IPv4\Address::factory('10.250.30.40');
		$this->assertEquals(10, $ip[-4]);
		$this->assertEquals(250, $ip[1]);
	}

	public function testArrayAccessWithOutOfRangeOctet()
	{
		$sut = IPv4\Address::factory('10.250.30.40');
		$this->expectException(InvalidArgumentException::class);
		$sut[4];
	}

	public function testArrayAccessThatOutOfRangeOctetIsNotSet()
	{
		$sut = IPv4\Address::factory('10.250.30.40');
		$this->assertFalse(isset($sut[4]));
	}

	public function testArrayAccessSet()
	{
		$this->expectException(\LogicException::class);
		$ip = IPv4\Address::factory('10.250.30.40');
		$ip[0] = 0;
	}

	public function testArrayAccessUnset()
	{
		$this->expectException(\LogicException::class);
		$ip = IPv4\Address::factory('10.250.30.40');
		unset($ip[0]);
	}
}


classes/Leth/IPAddress/IPv6/NetworkAddress.php
<?php
/*
 * This file is part of the PHP-IPAddress library.
 *
 * The PHP-IPAddress library is free software: you can redistribute it
 * and/or modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * The PHP-IPAddress library is distributed in the hope that it will
 * be useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with the PHP-IPAddress library.
 * If not, see <http://www.gnu.org/licenses/>.
 */
namespace Leth\IPAddress\IPv6;

use \Leth\IPAddress\IP;
use \Leth\IPAddress\IPv4;
use \Leth\IPAddress\IPv6;

class NetworkAddress extends \Leth\IPAddress\IP\NetworkAddress
{
	const IP_VERSION = 6;
	const MAX_SUBNET = 128;

	public static function generate_subnet_mask(int $subnet): IP\Address
	{
		$masks = [];
		for ($i=1; $i <= 4; $i++)
		{
			// left shift operates over arch-specific integer sizes,
			// so we have to special case 32 bit shifts
			$shift = min(32, max(0, 32*$i  - $subnet));
			if ($shift == 32)
			{
				$masks[] = 0;
			}
			else
			{
				$masks[] = (~0) << $shift;
			}
		}
		$result = unpack('H*', pack('N4', $masks[0], $masks[1], $masks[2], $masks[3]));
		return IPv6\Address::factory(implode(':', str_split($result[1], 4)));
	}

	/**
	 * Gets the Global subnet mask for this IP Protocol
	 *
	 * @author Marcus Cobden
	 */
	public static function get_global_netmask(): IP\Address
	{
		return static::generate_subnet_mask(static::MAX_SUBNET);
	}
}


classes/Leth/IPAddress/IP/NetworkAddressIterator.php
<?php
/*
 * This file is part of the PHP-IPAddress library.
 *
 * The PHP-IPAddress library is free software: you can redistribute it
 * and/or modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * The PHP-IPAddress library is distributed in the hope that it will
 * be useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with the PHP-IPAddress library.
 * If not, see <http://www.gnu.org/licenses/>.
 */
namespace Leth\IPAddress\IP;
use \Leth\IPAddress\IP;

class NetworkAddressIterator implements \Iterator
{
	/**
	 * The network of iterator
	 *
	 * @var IP\NetworkAddress
	 */
	protected $network;

	/**
	 * The position of iterator
	 *
	 * @var IP\Address
	 */
	protected $position;

	public function __construct(NetworkAddress $network)
	{
		$this->network = $network;
		$this->rewind();
	}

	/**
	 * Set the pointer of iterator to a first network address
	 */
	public function rewind() : void
	{
		$this->position = $this->network->get_network_start();
	}

	/**
	 * Get the value from iterator
	 */
	public function current(): IP\Address
	{
		return $this->position;
	}

	/**
	 * Get the key from iterator
	 */
	public function key(): string
	{
		return $this->position->__toString();
	}

	/**
	 * Move the pointer of iterator to a next network address
	 */
	public function next(): void
	{
		$this->position = $this->position->add(1);
	}

	/**
	 * Next network address is valid
	 */
	public function valid(): bool
	{
		return ($this->position->compare_to($this->network->get_network_end()) <= 0);
	}
}


classes/Leth/IPAddress/IPv6/Address.php
<?php
/*
 * This file is part of the PHP-IPAddress library.
 *
 * The PHP-IPAddress library is free software: you can redistribute it
 * and/or modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * The PHP-IPAddress library is distributed in the hope that it will
 * be useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with the PHP-IPAddress library.
 * If not, see <http://www.gnu.org/licenses/>.
 */
namespace Leth\IPAddress\IPv6;

use InvalidArgumentException;
use \Leth\IPAddress\IP, \Leth\IPAddress\IPv6, \Leth\IPAddress\IPv4;
use Math_BigInteger;

class Address extends IP\Address
{
	const IP_VERSION = 6;
	const FORMAT_ABBREVIATED = 2;
	const FORMAT_MAPPED_IPV4 = 3;
	// format mapped v4 if possible, else compact
	const FORMAT_MAY_MAPPED_COMPACT = 4;

	public static function factory(mixed $address): self
	{
		if ($address instanceof IPv6\Address)
		{
			return $address;
		}
		elseif (is_string($address))
		{
			if ( ! filter_var($address, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6))
			{
				throw new \InvalidArgumentException("'$address' is not a valid IPv6 Address.");
			}
			$address = inet_pton($address);
		}
		elseif ($address instanceOf Math_BigInteger)
		{
			// Do nothing
		}
		elseif (is_int($address))
		{
			$address = new Math_BigInteger($address);
		}
		else
		{
			throw new \InvalidArgumentException('Unsupported argument type.');
		}

		return new IPv6\Address($address);
	}

	/**
	 * This makes an IPv6 address fully qualified and padded.
	 * It replaces :: with appropriate 0000 blocks, and pads out all dropped 0s
	 *
	 * IE: 2001:630:d0:: becomes 2001:0630:00d0:0000:0000:0000:0000:0000
	 */
	public static function pad(string $address): string
	{
		$parts = explode(':', $address);
		$count = count($parts);

		$hextets = [];
		foreach ($parts as $i => $part)
		{
			if (isset($part[3])) // not need pad
			{
				$hextets[] = $part;
			}
			elseif ($part == '' && 0 < $i && $i < $count - 1) // missing hextets in ::
			{
				$missing = 8 - $count + 1;
				while ($missing--)
				{
					$hextets[] = '0000';
				}
			}
			else
			{
				$hextets[] = str_pad($part, 4, '0', STR_PAD_LEFT);
			}
		}

		return implode(':', $hextets);
	}

	protected function __construct($address)
	{
		if ($address instanceOf Math_BigInteger)
		{
			parent::__construct(str_pad($address->abs()->toBytes(), 16, chr(0), STR_PAD_LEFT));
		}
		else
		{
			parent::__construct($address);
		}
	}

	public function is_encoded_IPv4_address()
	{
		return strncmp($this->address, "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff", 12) === 0;
	}

	public function as_IPv4_address()
	{
		if(!$this->is_encoded_IPv4_address())
			throw new InvalidArgumentException('Not an IPv4 Address encoded in an IPv6 Address');
		list(,$hex) = unpack('H*', $this->address);
		$parts = array_map('hexdec', array_slice(str_split($hex, 2), 12));
		$address = implode('.', $parts);
		return IPv4\Address::factory($address);
	}

	public function add(int|Math_BigInteger $value): IP\Address
	{
		$left = new Math_BigInteger($this->address, 256);
		$right = ($value instanceof Math_BigInteger) ? $value : new Math_BigInteger($value);
		return new IPv6\Address($left->add($right));
	}

	public function subtract(int|Math_BigInteger $value): IP\Address
	{
		$left = new Math_BigInteger($this->address, 256);
		$right = ($value instanceof Math_BigInteger) ? $value : new Math_BigInteger($value);
		return new IPv6\Address($left->subtract($right));
	}

	/**
	  * Calculates the Bitwise & (AND) of a given IP address.
	  */
	  public function bitwise_and(IP\Address $other): IP\Address
	{
		return $this->bitwise_operation('&', $other);
	}

	/**
	  * Calculates the Bitwise | (OR) of a given IP address.
	  */
	  public function bitwise_or(IP\Address $other): IP\Address
	{
		return $this->bitwise_operation('|', $other);
	}

	/**
	  * Calculates the Bitwise ^ (XOR) of a given IP address.
	  */
	public function bitwise_xor(IP\Address $other): IP\Address
	{
		return $this->bitwise_operation('^', $other);
	}

	/**
	  * Calculates the Bitwise ~ (NOT) of a given IP address.
	  */
	public function bitwise_not(): IP\Address
	{
		return $this->bitwise_operation('~');
	}

	public function bitwise_operation($operation, $other = NULL)
	{
		if ($operation != '~')
		{
			$this->check_types($other);
		}

		switch ($operation) {
			case '&':
				$result = $other->address & $this->address;
				break;
			case '|':
				$result = $other->address | $this->address;
				break;
			case '^':
				$result = $other->address ^ $this->address;
				break;
			case '~':
				$result = ~$this->address;
				break;

			default:
				throw new \InvalidArgumentException('Unknown Operation type \''.$operation.'\'.');
				break;
		}

		return new IPv6\Address($result);
	}

	public function compare_to(IP\Address $other): int
	{
		$this->check_types($other);

		if ($this->address < $other->address)
			return -1;
		elseif ($this->address > $other->address)
			return 1;
		else
			return 0;
	}

	public function format($mode): string
	{
		list(,$hex) = unpack('H*', $this->address);
		$parts = str_split($hex, 4);

		if ($mode === IPv6\Address::FORMAT_MAY_MAPPED_COMPACT) {
			if ($this->is_encoded_IPv4_address()) {
				$mode = IPv6\Address::FORMAT_MAPPED_IPV4;
			} else {
				$mode = IPv6\Address::FORMAT_COMPACT;
			}
		}

		switch ($mode) {
			case IP\Address::FORMAT_FULL:
				// Do nothing
				break;

			case IPv6\Address::FORMAT_ABBREVIATED:
				foreach ($parts as $i => $quad)
				{
					$parts[$i] = ($quad === '0000') ? '0' : ltrim($quad, '0');
				}
				break;

			case IPv6\Address::FORMAT_MAPPED_IPV4:
				list($a, $b) = str_split($parts[6], 2);
				list($c, $d) = str_split($parts[7], 2);
				return '::ffff:' . implode('.', [hexdec($a), hexdec($b), hexdec($c), hexdec($d)]);
				break;

			case IP\Address::FORMAT_COMPACT:
				$best_pos   = $zeros_pos = FALSE;
				$best_count = $zeros_count = 0;
				foreach ($parts as $i => $quad)
				{
					$parts[$i] = ($quad === '0000') ? '0' : ltrim($quad, '0');

					if ($quad === '0000')
					{
						if ($zeros_pos === FALSE)
						{
							$zeros_pos = $i;
						}
						$zeros_count++;

						if ($zeros_count > $best_count)
						{
							$best_count = $zeros_count;
							$best_pos = $zeros_pos;
						}
					}
					else
					{
						$zeros_count = 0;
						$zeros_pos = FALSE;

						$parts[$i] = ltrim($quad, '0');
					}
				}


				if ($best_pos !== FALSE)
				{
					$insert = [NULL];

					if ($best_pos == 0 OR $best_pos + $best_count == 8)
					{
						$insert[] = NULL;
						if ($best_count == count($parts))
						{
							$best_count--;
						}
					}
					array_splice($parts, $best_pos, $best_count, $insert);
				}

				break;

			default:
				throw new \InvalidArgumentException('Unsupported format mode: '.$mode);
		}

		return implode(':', $parts);
	}

	public function __toString()
	{
		return $this->format(IPv6\Address::FORMAT_MAY_MAPPED_COMPACT);
	}
}
